src/operators/operators.h

```diff
@@ -263,21 +276,21 @@ class OperatorBase
     std::vector<std::vector<dealii::FullMatrix<real>>> basis_at_facet_cubature;
     ///Flux basis functions evaluated at facet cubature nodes.
     std::vector<std::vector<std::vector<dealii::FullMatrix<real>>>> flux_basis_at_facet_cubature;
```

Comment:
std::vector< std::vector<> > is EXTREMELY innefficient. A triple std::vector<std::vector<std::vector<>>> is even worse, and possibly going to consume all the memory available for even moderate-sized cases.

You data is all over the place in memory. The goal of using operators is to make the code more efficient. Using this data structure is almost certainly slower than recomputing things on the fly.

Reply:
Please do not store operators that aren't necessary and find a better way to store things contiguously.

This of course applies to all the various std::vector<std::vector<..........>> in here.

Furthermore, it's not clear which index corresponds to what (face, cubature node, element)

Reply:
One good test case is to add a large 3D test case with over 5 million degrees of freedom, and attempt to initialize this. The test case itself can simply be about allocating it and make sure it doesn't crash.

Reply:
I don't see how these will blow up for a 5 million dof test. They are all local operators, for example, it only stores one local mass matrix for the corresponding polynomial degree. None of the operators are global. Also, from the way I implemented the sum-factorization, we will most likely only need to use the 1D version of the operators class. Also, from experience, do you have a suggestion to replace the std::vector<std::vector<> > ? I initially wanted to use array but the sizes aren't known at declaration.

Reply:
Yeah that's fine, then a 5 million DoFs shouldn't make it blow up? Best way to find out and make sure it doesn't is to add the test case. What we don't want is to merge this code in, and all of a sudden you can't initialize the strong form with 5 million DoFs.

Firstly, in some of your array sizes, some array dimensions are using dim, or number of faces, which depends on the number of dimensions which are known at compile time.

Second, the main reason you don't want to use std::vector<std::vector<...>> is that it's not stored contiguously. This will be important for data that you are accessing in a sequence rapidly. For example, one cell's DoFs should be stored contiguously. Then its neighbor cell should be stored nearby in memory if possible. Please read the "Detailed description" from https://www.dealii.org/developer/doxygen/deal.II/classTableBase.html. When you have a std::vector<std::vector<>> or std::vector<Table>, you are actually storing a vector of pointers to other vectors or a vector of pointers to other tables. It is therefore slow to access because myVec[0][] is not near myVec[1][]. _Sometimes_, this is OK if you know you won't be accessing myVec[1][] soon after myVec[0][] (maybe two different grid orders). However, using nested std::vector blindly is a redflag that needs proper analysis before it makes it through.

Third, I'm noticing that you are basically using the OperatorBase to basically store data as if they are global variables. For example `gradient_mapping_shape_functions_face_flux_nodes`, seems to only be used within operators.cpp as an intermediary variable, there is no point in storing those.

Reply:
Thank you very much for your detailed response and clarifying the issue with contiguous storing! I will clean out the std::vectors as much as possible. I will probably leave the vectors for storing polynomial degrees, but will use your suggestion of arrays for dim and n_faces. Also, with regards to gradient_mapping_shape_functions_face_flux_nodes, it is used every time we compute a metric term on the face, that it why it is being stored.

Reply:
So locally I've come across a problem: I've made a operators factory and a derived OperatorsBaseState class templated on nstate and n_faces to start introducing std::arrays for them (and dim), but my issue is that all the operators needed for dg etc are members of the derived class, whereas in dg, we can only store an object of the base class (since nstate isn't a compile time constant at the DGBase level). I've tried using dynamic_cast (the only solution I've been able to find online) to no success because of this very issue. My current solution is to store some operators in the base (the ones not dependent on nstate) and others in the state (derived) class, but then in DGStrong won't I have to store an object of the derived class? I feel like this is having me store 2 objects, one base in DGBase and one derived in DGStrong. Is there a way to reduce the memory cost, by, for example, having DGBase use a derived object stored in DGStrong, or some other suggestion?

Reply:
No need to template on n_faces if you already have templated dim.

Both DGStrong and DGWeak are derived from DGBaseState. Note that if you have a factory of OperatorsBase, you should return an std::unique_ptr<OperatorBase>, which then doesn't depend on the state. That can be given to DGBase. For example, ODESolver has a DGBase, which doesn't depend on the state even though in the implementation, there are state variables.

Reply:
Yes that is what I did.
So the unique pointer returns a OperatorsBase not dependent on the state. The OperatorsBase has a derived class called OperatorsBaseState that depends on the state.
This derived state class is needed in DGStrong and DGWeak (DGBaseState).
The issue isn't DGBase or DGBaseState having an OperatorsBase, the issue is that the DGBaseState/DGStrong/DGWeak also needs an OperatorsBaseState to use state dependent operators. This is making me store an OperatorsBase object in DGBase, and an additional OperatorsBaseState object in DGBaseState/DGStrong/DGWeak.
My question was more wondering if there was a better solution to be able to use the OperatorsBaseState derived class in DGBaseState/DGStrong/DGWeak while only storing the OperatorsBase in DGBase? This is because the base object can't access the derived functions/operators.

Reply:
Aaah, so what is the point of having OperatorBase and OperatorBaseState?

The main goal inheritance is to have an _interface to have polymorphic behavior_. You would usually want to have various Operators derived from OperatorBaseState. For example: https://www.dealii.org/developer/doxygen/deal.II/classFiniteElement.html
has an interface for various functions like `shape_value`, `shape_grad`, etc.
This is only useful is you plan on swapping around your implementation by having other derived classes, which you don't seem to have? Are you able to come up with another class that you would derive from the same OperatorBase? If not, then you probably don't need to use inheritance here.

The other case for inheritance that can be helpful is for the [template method pattern](https://en.wikipedia.org/wiki/Template_method_pattern) where you reuse the sum factorization implementation of the parent class for various different operators. Right now, your OperatorBase or OperatorBaseState is not actually a "useful" class. It is the equivalent of storing everything globally in a namespace.

An example of template method pattern you can use for Operators is:
```cpp
template<int dim, typename real>
class OperatorBase<real> {
    // virtual function to be defined
    void applyOperator(const std::vector<real> input_vector, const std::vector<real> output_vector) = 0;
}
template<int dim, typename real>
class SumFactorizedOperator : public OperatorBase<real> 

public:
    // Precompute 1D operator in constructor
    SumFactorizedOperator() : oneD_operator(build1DOperator()) { };

    void applyOperator(const std::vector<real> input_vector, const std::vector<real> output_vector) override const {
        // Code using apply1DOperator(input_vector, output_vector)
    }

private:
    void apply1DOperator(const std::vector<real> input_vector, const std::vector<real> output_vector) const {
        oneD_operator * input_vector;
    }

protected:
    virtual dealii::FullMatrix<real> build1DOperator() = 0 { // code to build it
    }
    dealii::FullMatrix<real> oneD_operator;
}

template<int dim, typename real, typename >
class LocalMassMatrix : public SumFactorizedOperator {

protected:
    dealii::FullMatrix<real> build1DOperator() override { 
         // Basically the only thing that changes from one SumFactorizedOperator to another.
    }
}
template<int dim, typename real, typename >
class LocalStiffnessMatrix : public SumFactorizedOperator {

protected:
    dealii::FullMatrix<real> build1DOperator() override { 
         // Basically the only thing that changes from one SumFactorizedOperator to another.
    }
}
```
Of course, the above can also be templated over `nstate` if need be. DG would then hold those operators directly instead of their interface. That is, DG would directly hold the mass and stiffness matrices like:
```cpp
// in dg.h
class DGBaseState : public DGBase
{
    LocalMassMatrix local_mass_matrix;
    std::array<LocalStiffnessMatrix, dim>  stiffness_matrices;
}
```
All the above is similar to:
https://www.dealii.org/developer/doxygen/deal.II/classMatrixFreeOperators_1_1Base.html

OR
You could just not have any derived classes. Just have a different class for MassMatrix, a class for StiffnessMatrix, etc. You don't have to have inheritance. Maybe those MassMatrix, StiffnessMatrix call some similar matrix-vector function. However, since you want to cache your operators after computing them once, you can't get away from creating classes to have a neat location to store your data. You just don't want to store "all of it" in the same class.

Reply:
Ok thank you. I think for now I will just leave it as a std::vector of nstate and in the future add the template to make it an std::array by nstate.

Reply:
No no! regardless of the class structure described above, having compile-time constant for nstate is important!

It is *so* much harder to add templated parameters later on.

Reply:
Ok, so my current plan is to have the two forms, OperatorsBase and OperatorsBaseState as described earlier to template the arrays by state. DGBase will only hold the Base, while DGBaseState will hold the other. Then, after that works, I will follow your suggestion, either directly in this PR or opening an issues ticket, by making each operator itself a derived class on both basic and sumfactorization forms. 
Thank you very much for your detailed explanation above, it's helped me out a lot with understanding the role of inheritance and polymorphic behaviour.

Reply:
fixed